### PR TITLE
Round-robin over worker nodes for 0-shard router queries

### PR DESCRIPTION
**Pull request by marcocitus**
When a router query prunes down to 0 shards, we generate a dummy placement which represents an arbitrary worker node (the first primary in `pg_dist_node`). Given that we'll start using this feature to plan queries on intermediate result files, we want to distribute the load across workers using round-robin.

### CODE REVIEW

File: src/backend/distributed/planner/multi_router_planner.c [hunk 1]
```diff
@@ -1629,6 +1629,8 @@ PlanRouterQuery(Query *originalQuery, RelationRestrictionContext *restrictionCon
 				List **placementList, uint64 *anchorShardId, List **relationShardList,
 				bool replacePrunedQueryWithDummy, bool *multiShardModifyQuery)
 {
+	static int zeroShardQueryRoundRobin = 0;
```
Comment (metdos) [inline src/backend/distributed/planner/multi_router_planner.c]: Can we have overflow problem if the server lives long enough?

Comment (pykello) [reply src/backend/distributed/planner/multi_router_planner.c]: Yeah, I think this at least should be unsigned so `%` doesn't return negative index when overflow happens.

Comment (marcocitus) [reply src/backend/distributed/planner/multi_router_planner.c]: Yep, good catch! Fixed it.

**Comment by metdos** [comment]
It seems you also hit pg_dist_transaction problem :)
```
 SELECT count(*) FROM pg_dist_transaction;
  count 
 -------
-     4
+     6
 (1 row)
```

**Comment by marcocitus** [comment]
> It seems you also hit pg_dist_transaction problem :)

yeh, we shouldn't be running `multi_mx_transaction_recovery` in parallel with `multi_mx_modifications`.

**Comment by marcocitus** [comment]
Fixed up the master branch and rebased.

**Comment by metdos** [comment]
I think you can 🚢

**Pull request merged**